ui/signed-out/src/commonMain/kotlin/social/androiddev/signedout/signin/SignInViewModel.kt

```diff
@@ -9,49 +9,116 @@
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import social.androiddev.domain.authentication.model.ApplicationOAuthToken
+import social.androiddev.domain.authentication.usecase.CreateAccessToken
+import social.androiddev.domain.authentication.usecase.GetSelectedApplicationOAuthToken
+import social.androiddev.signedout.util.encode
 import java.net.URI
+import kotlin.coroutines.CoroutineContext
 
-// TODO USE DI
-// Add tests
-class SignInViewModel(
-    private val server: String
-) {
-
-    fun getSignInUrl(): String {
-        // TODO
-        //  Use the right client id and redirect url and get them from a secure location
-        //  build url with an url builder my be using Ktor !!
-        return "https://$server/oauth/authorize?client_id=$CLIENT_ID&scope=$OAUTH_SCOPES&redirect_uri=https://dodomastodon.com&response_type=code"
+// TODO Add tests
+internal class SignInViewModel(
+    mainContext: CoroutineContext,
+    private val getSelectedApplicationOAuthToken: GetSelectedApplicationOAuthToken,
+    private val createAccessToken: CreateAccessToken,
+    private val onSignedIn: () -> Unit,
+) : InstanceKeeper.Instance {
+
+    // The scope survives Android configuration changes
+    private val scope = CoroutineScope(mainContext + SupervisorJob())
+
+    private val _state = MutableStateFlow(createInitialState())
+    val state: StateFlow<SignInComponent.State> = _state.asStateFlow()
+
+    private fun createInitialState() =
+        SignInComponent.State(oauthAuthorizeUrl = "", redirectUri = "", server = "")
+
+    init {
+        scope.launch {
+            val token = getSelectedApplicationOAuthToken()
+            _state.update {
+                it.copy(
+                    server = token.server,
+                    redirectUri = token.redirectUri,
+                    oauthAuthorizeUrl = createOAuthAuthorizeUrl(token),
+                )
+            }
+        }
     }
 
-    fun resolveSignInStatusFromUrl(
-        url: String,
-        onSignedIn: () -> Unit,
-        onFailed: (error: String) -> Unit
-    ) {
+    private fun createOAuthAuthorizeUrl(token: ApplicationOAuthToken): String {
+        val b = StringBuilder().apply {
+            append("https://${token.server}")
+            append("/oauth/authorize?client_id=${token.clientId}")
+            append("&scope=${"read write follow push".encode()}")
+            append("&redirect_uri=${token.redirectUri.encode()}")
+            append("&response_type=code")
+        }
+        return b.toString()
+    }
+
+    fun onErrorFromOAuth(error: String) {
+        displayErrorWithDuration(error)
+    }
+
+    private fun displayErrorWithDuration(error: String) {
+        _state.update { it.copy(error = error) }
+        scope.launch {
+            delay(3_000)
+            _state.update { it.copy(error = null) }
+        }
+    }
+
+    fun shouldCancelLoadingUrl(url: String): Boolean {
```

Comment:
I don't think this is a blocking point for this PR, but maybe we must enhance the extraction of query params from the URL,  because the current implementation does not support the reading for serval params at the same time, And Based on the value of those params :

1. `DONE` If we have a `code` query param, we navigate to the home ( it is already implemented, and that works fine)

2. If we have an error code param :

 - `TODO` If the error value is `access_denied`, we must go back to the selection server screen (for now, the user stays at the login screen and loged into his mastodon account from the app login webview). That happens when the user denies the authorization requested by Dodo app.
 
-  `TODO` If the error value is other than `access_denied`, we must read the `error_description` param to display its content to the user.

Reply:
Sounds good - I did see a TODO you left here previously. Can you please follow-up on this in a different PR? It's not as high priority so no rush

Reply:
created https://github.com/AndroidDev-social/DodoForMastodon/issues/96